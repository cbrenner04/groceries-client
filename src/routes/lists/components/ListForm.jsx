import React, { useState } from 'react';
import PropTypes from 'prop-types';

import { SelectField, TextField } from '../../../components/FormFields';

function ListForm(props) {
  const defaultListType = 'GroceryList';
  const [name, setName] = useState('');
  const [type, setType] = useState(defaultListType);

  const handleSubmit = (event) => {
    event.preventDefault();
    props.onFormSubmit({
      name,
      type,
    });
    setName('');
    setType(defaultListType);
  };

  return (
    <form className="form" onSubmit={handleSubmit} autoComplete="off">
      <TextField
        name="name"
        label="Name"
        value={name}
        handleChange={({ target: { value } }) => setName(value)}
        placeholder="My super cool list"
      />
      <SelectField
        name="type"
        label="Type"
        value={type}
        handleChange={({ target: { value } }) => setType(value)}
        options={[
          { value: 'BookList', label: 'books' },
          { value: 'GroceryList', label: 'groceries' },
          { value: 'MusicList', label: 'music' },
          { value: 'ToDoList', label: 'to-do' },
        ]}
        blankOption={false}
      />
      <button type="submit" className="btn btn-success btn-block">Create List</button>
    </form>
  );
}

ListForm.propTypes = {
  onFormSubmit: PropTypes.func.isRequired,
};

export default ListForm;
