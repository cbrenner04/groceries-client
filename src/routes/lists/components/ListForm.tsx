import type { ChangeEvent, FormEvent } from 'react';
import React, { useState } from 'react';
import { Button, Collapse, Form } from 'react-bootstrap';

import ListFormFields from '../components/ListFormFields';
import FormSubmission from '../../../components/FormSubmission';
import type { IList } from '../../../typings';
import { EListType } from '../../../typings';

interface IListFormProps {
  onFormSubmit: (list: IList) => Promise<void>;
  pending: boolean;
}

const ListForm: React.FC<IListFormProps> = ({ onFormSubmit, pending }) => {
  const defaultListType = EListType.GROCERY_LIST;
  const [name, setName] = useState('');
  const [type, setType] = useState(defaultListType as EListType);
  const [showForm, setShowForm] = useState(false);

  const handleSubmit = async (event: FormEvent<HTMLFormElement>) => {
    event.preventDefault();
    // TODO: figure this out.
    await onFormSubmit({ name, type } as IList);
    setName('');
    setType(defaultListType);
  };

  return (
    <>
      {!showForm && (
        <Button variant="link" onClick={() => setShowForm(true)} aria-controls="form-collapse" aria-expanded={showForm}>
          Add List
        </Button>
      )}
      <Collapse in={showForm}>
        <Form id="form-collapse" onSubmit={handleSubmit} autoComplete="off">
          <ListFormFields
            name={name}
            type={type}
            handleNameChange={({ target: { value } }: ChangeEvent<HTMLInputElement>) => setName(value)}
            handleTypeChange={({ target: { value } }: ChangeEvent<HTMLInputElement>) => setType(value as EListType)}
          />
          <FormSubmission
            disabled={pending}
            submitText="Create List"
            cancelAction={() => setShowForm(false)}
            cancelText="Collapse Form"
            displayCancelButton={true}
          />
        </Form>
      </Collapse>
    </>
  );
};

export default ListForm;
