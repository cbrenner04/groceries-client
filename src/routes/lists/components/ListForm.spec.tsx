import React from 'react';
import { render, type RenderResult, waitFor } from '@testing-library/react';
import userEvent, { type UserEvent } from '@testing-library/user-event';

import ListForm, { type IListFormProps } from './ListForm';

interface ISetupReturn extends RenderResult {
  user: UserEvent;
  props: IListFormProps;
}

function setup(suppliedProps?: Partial<IListFormProps>): ISetupReturn {
  const user = userEvent.setup();
  const defaultProps = {
    pending: false,
    onFormSubmit: jest.fn(),
  };
  const props = { ...defaultProps, ...suppliedProps };
  const component = render(<ListForm {...props} />);

  return { ...component, props, user };
}

describe('ListForm', () => {
  it('renders', () => {
    const { container } = setup();

    expect(container).toMatchSnapshot();
  });

  it('expands form', async () => {
    const { findByText, queryByLabelText, user } = setup();

    expect(queryByLabelText('Name')).not.toBeInTheDocument();
    await user.click(await findByText('Add List'));
    expect(await findByText('Collapse Form')).toBeInTheDocument();
    expect(await findByText('Create List')).toBeInTheDocument();
    expect(await findByText('Type')).toBeInTheDocument();
    expect(await findByText('Name')).toBeInTheDocument();
  });

  it('collapses form', async () => {
    const { findByText, queryByLabelText, user } = setup();

    await user.click(await findByText('Add List'));
    expect(await findByText('Collapse Form')).toBeInTheDocument();
    await user.click(await findByText('Collapse Form'));
    await waitFor(() => {
      expect(queryByLabelText('Name')).not.toBeInTheDocument();
    });
  });

  it('changes the value in the name field', async () => {
    const { findByLabelText, findByText, user } = setup();

    await user.click(await findByText('Add List'));
    await user.type(await findByLabelText('Name'), 'foo');

    expect(await findByLabelText('Name')).toHaveValue('foo');
  });

  it('changes the value in the type field', async () => {
    const { findByLabelText, findByText, user } = setup();

    await user.click(await findByText('Add List'));
    await user.selectOptions(await findByLabelText('Type'), 'MusicList');

    expect(await findByLabelText('Type')).toHaveValue('MusicList');
  });

  it('calls props.onFormSubmit when form is submitted', async () => {
    const onFormSubmit = jest.fn().mockResolvedValue({});
    const { findByLabelText, findAllByRole, findByText, props, user } = setup({ onFormSubmit });

    await user.click(await findByText('Add List'));
    await user.type(await findByLabelText('Name'), 'foo');
    await user.selectOptions(await findByLabelText('Type'), 'BookList');
    await user.click((await findAllByRole('button'))[0]); // 'Create List' is now the first button

    await waitFor(() => expect(props.onFormSubmit).toHaveBeenCalledTimes(1));

    expect(onFormSubmit).toHaveBeenCalledWith({ name: 'foo', type: 'BookList' });
  });

  it('disables submit when in pending state', async () => {
    const onFormSubmit = jest.fn().mockResolvedValue({});
    const { findByText, findAllByRole } = setup({ pending: true, onFormSubmit });

    await userEvent.click(await findByText('Add List'));
    expect((await findAllByRole('button'))[0]).toBeDisabled();
  });
});
