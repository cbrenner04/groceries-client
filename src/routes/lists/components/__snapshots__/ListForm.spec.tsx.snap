// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ListForm renders 1`] = `
<div>
  <button
    aria-controls="form-collapse"
    aria-expanded="false"
    class="btn btn-link"
    type="button"
  >
    Add List
  </button>
</div>
`;
